library/net.cpp

```diff
@@ -1410,3 +1415,276 @@ int dnet_trans_forward(struct dnet_io_req *r, struct dnet_net_state *orig, struc
+		if (!r)
+			throw std::bad_alloc();
+
+		r->io_req_type = DNET_IO_REQ_TYPED_RESPONSE;
+		r->response_info = new n2_response_info({ cmd, std::move(response_holder) });
+
+		r->st = dnet_state_get(st);
+		dnet_schedule_io(st->n, r);
+	};
+
+	n2_repliers repliers_wrappers;
+	repliers_wrappers.reply_error = [reply_error = std::move(repliers.reply_error), enqueue_response](int errc){
+		enqueue_response(std::bind(reply_error, errc));
+	};
+	repliers_wrappers.reply = [reply = std::move(repliers.reply),
+	                           enqueue_response](std::unique_ptr<n2_message> msg){
+		enqueue_response(n2_bind_noncopyable(reply, std::move(msg)));
+	};
+	return repliers_wrappers;
+}
+
+static int n2_trans_send(dnet_trans *t, n2_request_info *request_info) {
+	using namespace ioremap::elliptics;
+
+	struct dnet_net_state *st = t->st;
+	struct dnet_test_settings test_settings;
+	int err;
+
+	dnet_trans_get(t);
+
+	BOOST_SCOPE_EXIT(&t) {
+		dnet_trans_put(t);
+	} BOOST_SCOPE_EXIT_END
+
+	pthread_mutex_lock(&st->trans_lock);
+	err = dnet_trans_insert_nolock(st, t);
+	if (!err) {
+		dnet_trans_update_timestamp(t);
+		dnet_trans_insert_timer_nolock(st, t);
+	}
+	pthread_mutex_unlock(&st->trans_lock);
+	if (err)
+		return err;
+
+	if (t->n->test_settings && !dnet_node_get_test_settings(t->n, &test_settings) &&
+	    test_settings.commands_mask & (1 << t->command))
+		return err;
+
+	auto repliers_wrappers = n2_make_repliers_via_request_queue(st,
+	                                                            request_info->request->cmd,
+	                                                            std::move(request_info->repliers));
+
+	n2::net_state_get_protocol(st)->send_request(st,
+	                                             std::move(request_info->request),
+	                                             std::move(repliers_wrappers));
+
+	return err;
+}
+
+static int n2_trans_complete_forward(struct dnet_addr * /*addr*/, n2_response_info *response_info, void *priv) {
+	auto cmd = &response_info->cmd;
+
+	if (!is_trans_destroyed(cmd)) {
+		response_info->response_holder();
+		return 0;
+	}
+
+	return -EINVAL;
+}
+
+int n2_trans_forward(n2_request_info *request_info, struct dnet_net_state *orig, struct dnet_net_state *forward) {
+	dnet_cmd *cmd = &request_info->request->cmd;
+
+	auto t = dnet_trans_alloc(orig->n, 0);
+	if (!t)
+		return -ENOMEM;
+
+	t->rcv_trans = cmd->trans;
+	cmd->trans = t->cmd.trans = t->trans = atomic_inc(&orig->n->trans);
+
+	t->cmd = *cmd;
+
+	t->wait_ts = [&cmd, &t, &request_info]() -> timespec {
+		using namespace ioremap::elliptics;
+
+		dnet_time deadline = request_info->request->deadline;
+
+		if (dnet_time_is_empty(&deadline)) {
+			return t->wait_ts;
+		}
+
+		dnet_time current;
+		dnet_current_time(&current);
+		if (dnet_time_before(&deadline, &current)) {
+			return timespec{0, 0};
+		}
+
+		auto diff = nanosec_to_dnet_time(dnet_time_to_nanosec(deadline) - dnet_time_to_nanosec(current));
```

Comment:
Yes I can, but `dnet_trans_forward` is legacy: as I understood, in elliptics is strict politic to beware of touching legacy.

I thought about writing `operator`s (+, -, ..) for `dnet_time`, but throwed out this idea, because my opinion is constant: `dnet_time` is evil concept, and more evil to use it in internal calculations. I want not support and upgrade this evil.

Reply:
Ok, nanosec_to_dnet_time & dnet_time_to_nanosec will be deleted.

Reply:
They're deleted.